### PR TITLE
Transparent fallback for BeautifulSoup lxml -> html5lib -> html.parser

### PR DESCRIPTION
**Pull request by pseyfert**
ran into problems getting down-frab-videos running. turned out the easiest fix is using html5lib as parser.

 * Added an error message giving the hint
 * Added a config switch

What do you think?

### CODE REVIEW

File: down_frab_videos.py [hunk 1]
```diff
@@ -254,14 +279,15 @@ def get_format_list(media_prefix):
             'From': SOURCE,
         }
 
+        print("getting ", media_prefix + "/")
```
Comment (mfherbst) [inline down_frab_videos.py]: Any reason why you added this print statement?

**Review by mfherbst — COMMENTED** [review]
Thanks for the PR.

It feels to me like a very hacky way to solve this. Is it not maybe better to use one of them by default and fall back to the other automatically if there are problems?

I mean I do not see the point, why this should be exposed to the config at all.

**Comment by mfherbst** [comment]
**Update**: This comment is obsolete. Read on to the next ;)

~~Also I find it really funny you are getting this error. Because I never did on any of my systems. Any clue why?~~

~~Would it maybe make sense to have `html5lib` as the default and not have `lxml` at all? Would that solve your problem?~~

**Comment by mfherbst** [comment]
Ok. I had a look at [the BeatifulSoup docs](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#installing-a-parser). From that I would suggest we have an automatic fallback mechanism `lxml`, `html5lib`, `html.parser`. If `lxml` cannot be found, I think we should issue a warning to suggest installation of the `lxml` python package.

What do you think?

File: down_frab_videos.py (hunk 1 above)
Comment (pseyfert) [reply down_frab_videos.py]: whoops. overlooked when cleaning up after debugging.

**Comment by mfherbst** [comment]
Maybe what I suggested above is a bit over the top after all. Does installing `lxml` solve your problem ... because if it does, than I would just bail out and issue an appropriate message, instructing to install the package.

File: down_frab_videos.py (hunk 1 above)
Comment (mfherbst) [reply down_frab_videos.py]: Thought so ;).

**Comment by pseyfert** [comment]
> It feels to me like a very hacky way to solve this.

Of cause it's hacky, I wanted to get the download started ;)

> I mean I do not see the point, why this should be exposed to the config at all.

Well, that's why I was asking for feedback. My thinking was that testing and
falling back once per video instead of once per installation doesn't seem
elegant. But either way is fine with me.

> If lxml cannot be found, I think we should issue a warning to suggest installation of the lxml python package.

That's the part I haven't understood so far about my failure, `lxml` is
installed, not sure why `BeautifulSoup` doesn't manage to pick it up. (I'm
undecided how dig I want to dig in understanding it.)

**Comment by mfherbst** [comment]
> Of cause it's hacky, I wanted to get the download started ;)

I understand. Sorry for the inconvenience ;).

> My thinking was that testing and
falling back once per video instead of once per installation doesn't seem
elegant.

It does not happen once per video. BeautifulSoup is only needed at the initialisation stage for parsing the media files and the media formats, so I think having a built-in fallback, which is walked through each time is ok. And well ... due to the oddities and irregularities in the way the VOC does the language codes and so on the code is not really in the cleanest stage either.

> That's the part I haven't understood so far about my failure, lxml is
installed, not sure why BeautifulSoup doesn't manage to pick it up.

Strange. Well, I guess the automatic fall back works for both of us and we do not have to dig too much ;).

File: down_frab_videos.py [hunk 2]
```diff
@@ -241,6 +243,28 @@ def __init__(self, short_message, long_message):
...
+        return BeautifulSoup(content, "html.parser")
+    except bs4.FeatureNotFound:
+        print("Warning: could not parse with 'html.parser', check your installation. "
+              "Falling back to 'html5lib'.")
+
+    try:
+        return BeautifulSoup(content, "html5lib")
+    except bs4.FeatureNotFound:
+        print("Error: could not parse with any of 'lxml', 'html.parser' and 'html5lib'. "
+              "Giving up.")
```
Comment (mfherbst) [inline down_frab_videos.py]: From the BeautifulSoup docs, I would say that `html.parser` is the worst and should be used last, right?

File: down_frab_videos.py [hunk 3]
```diff
@@ -241,6 +243,28 @@ def __init__(self, short_message, long_message):
         self.long_message = long_message
 
 
+def wrap_bs4(content):
```
Comment (mfherbst) [inline down_frab_videos.py:246]: You could write this routine as a `for`-loop over a list `["lxml", "html.parser", "html5lib"]`. This makes it easer to extend the list at a later point.

**Review by mfherbst — CHANGES_REQUESTED** [review]
Thanks for adapting. Apart from the order, just nits.

File: down_frab_videos.py (hunk 3 above)
Comment (pseyfert) [reply down_frab_videos.py:246]: good point

File: down_frab_videos.py (hunk 2 above)
Comment (pseyfert) [reply down_frab_videos.py]: Depends on your figure of merit. But let's put `html.parser` last.

File: down_frab_videos.py [hunk 4]
```diff
@@ -244,23 +244,18 @@ def __init__(self, short_message, long_message):
...
+    """
+    known_parsers = ["lxml", "html5lib", "html.parser"]
+    for parser in known_parsers:
+        try:
+            return BeautifulSoup(content, parser)
+        except bs4.FeatureNotFound:
+            print("Warning: could not parse with {}, "
+                  "check your installation. ".format(parser))
+            if parser != known_parsers[-1]:
+                print("Falling back to next known parser.")
```
Comment (mfherbst) [inline down_frab_videos.py:258]: Would be nice to know which one it tried unsuccessfully.

**Review by mfherbst — APPROVED** [review]
Great work.

File: down_frab_videos.py (hunk 4 above)
Comment (pseyfert) [reply down_frab_videos.py:258]: Maybe I don't understand, maybe you overlooked. That's the `{}` in line 255. For me it shows:

```
Warning: could not parse with lxml, check your installation. 
Falling back to next known parser.
```

Comment (mfherbst) [reply down_frab_videos.py:258]: I overlooked.

**Pull request merged**